### PR TITLE
Pass shape infos to ONNX -> Caffe2 C++ conversion backend

### PR DESCRIPTION
**Pull request by bddppq**
And let Gemm conversion to inspect the input `C` to try converting to FC.

### CODE REVIEW

File: caffe2/onnx/backend.h [hunk 1]
```diff
@@ -11,6 +11,8 @@
 #include <unordered_map>
 #include <unordered_set>
 
+#define kKnownOpsetVersion 6
```
Comment (yinghai) [inline caffe2/onnx/backend.h]: Constexpr?

File: caffe2/onnx/backend.h [hunk 2]
```diff
@@ -19,6 +21,25 @@ using ::ONNX_NAMESPACE::GraphProto;
 using ::ONNX_NAMESPACE::ModelProto;
 using ::ONNX_NAMESPACE::NodeProto;
 using ::ONNX_NAMESPACE::TensorProto;
+using ::ONNX_NAMESPACE::ValueInfoProto;
+
+using ValueInfosMap = std::unordered_map<std::string, ValueInfoProto>;
```
Comment (yinghai) [inline caffe2/onnx/backend.h]: Nit: ValueInfoMap

File: caffe2/python/pybind_state.cc [hunk 3]
```diff
@@ -793,13 +793,20 @@ void addObjectMethods(py::module& m) {
...
              int opset_version) -> std::vector<std::vector<py::bytes>> {
             // Note that we return two lists of serialized ops. The first set is
             // init_ops and the second set is ops for pred net. When converting
             // RNN related op, it is possible that we will create ops in the
             // init_net. Hence the return structure here
+            caffe2::onnx::ValueInfosMap value_infos{};
+            for (const auto& vi_bytes : value_infos_bytes) {
+              ::ONNX_NAMESPACE::ValueInfoProto vi{};
+              vi.ParseFromString(vi_bytes);
+              value_infos[vi.name()] = std::move(vi);
```
Comment (yinghai) [inline caffe2/python/pybind_state.cc]: Use emplace?

**Review by yinghai — APPROVED** [review]
LG. Some nits.

File: caffe2/python/onnx/backend.py [hunk 4]
```diff
@@ -708,90 +709,53 @@ def prepare(cls, model, device='CPU', raw_values_dict=None, **kwargs):
...
         if cbackend.support_onnx_import(node_def.op_type):
-            op_strs = cbackend.convert_node(node_def.SerializeToString(), opset_version)
+
+            # extract value infos from pred model (value infos of
+            # node's inputs that are in init model should be all
+            # available in pred model)
+            value_infos = []
+            for name in node_def.input:
+                if pred_model is not None:
+                    for vi in itertools.chain(pred_model.graph.input, pred_model.graph.output):
```
Comment (houseroad) [inline caffe2/python/onnx/backend.py]: Also add pred_model.graph.value_info?

**Review by houseroad — APPROVED** [review]
LGTM, thanks!

**Pull request closed without merging**